### DISCUSSION TITLE
Proper usage under slow internet?

### DISCUSSION THREAD

**Discussion by 1-1-2** [abraunegg/onedrive#2350] [Q&A] [answered]
ENV
---
Version: v2.4.23-1+np1
Environment: Debian 11

Description
---
I'm using onedrive in `--monitor` mode(conf attach below), in order to backup all files in my onedrive space to my own NAS.

```ini
sync_dir_permissions = "750"
sync_file_permissions = "640"
disable_notifications = "true"
local_first = "false"
no_remote_delete = "false"
download_only = "true"
upload_only = "false"
skip_dotfiles = "false"
monitor_interval = "60"
monitor_fullscan_frequency = "12"
rate_limit = "0"
```

Occasionally, I got some error message like:
<details>
<summary>piece of log</summary>
<pre>
```
2023-Mar-23 06:19:17.6280422	Downloading file path/to/a.docx ... done.
2023-Mar-23 07:19:18.1794674	ERROR: Microsoft OneDrive API returned an error with the following message:
2023-Mar-23 07:19:18.1796	  Error Message:    Timeout was reached on handle 559773A0FDE0
2023-Mar-23 07:19:21.7056616	ERROR: File download size mis-match. Increase logging verbosity to determine why.
2023-Mar-23 07:19:21.7057049	ERROR: File download hash mis-match. Increase logging verbosity to determine why.
2023-Mar-23 07:19:21.7057193	INFO: Potentially add --disable-download-validation to work around this issue but downloaded data integrity cannot be guaranteed.
2023-Mar-23 08:19:24.4225341	ERROR: Microsoft OneDrive API returned an error with the following message:
2023-Mar-23 08:19:24.4226691	  Error Message:    Timeout was reached on handle 559773A0FDE0
2023-Mar-23 08:19:28.0498736	ERROR: File download size mis-match. Increase logging verbosity to determine why.
2023-Mar-23 08:19:28.0499211	ERROR: File download hash mis-match. Increase logging verbosity to determine why.
2023-Mar-23 08:19:28.0499344	INFO: Potentially add --disable-download-validation to work around this issue but downloaded data integrity cannot be guaranteed.
2023-Mar-23 08:19:28.2619394	Sync with OneDrive is complete
2023-Mar-23 08:20:28.6380878	Starting a sync with OneDrive
2023-Mar-23 08:20:32.2456981	Sync with OneDrive is complete
2023-Mar-23 08:21:32.2543663	Starting a sync with OneDrive
2023-Mar-23 08:21:34.4909249	Sync with OneDrive is complete
2023-Mar-23 08:22:34.4997436	Starting a sync with OneDrive
2023-Mar-23 08:22:36.0365518	Sync with OneDrive is complete
2023-Mar-23 08:23:36.0449065	Starting a sync with OneDrive
2023-Mar-23 08:23:37.5515675	Sync with OneDrive is complete
2023-Mar-23 08:24:37.5601787	Starting a sync with OneDrive
2023-Mar-23 08:24:37.5603126	Syncing changes and items from OneDrive ...
.
[Duplicate content]
.
2023-Mar-23 09:51:09.2717882	Sync with OneDrive is complete
```
</pre></details>

With some log research, I found out this happend while downloading a large file (like 12.34g) with slow connection rate (like 300kb/s), which took hours to finish. After the errors, these file will not be sync any more even after `'monitor_interval' x 'monitor_fullscan_frequency'` cycle time. To solve this, I have to do a `--resync`.

Question
---

I've read those article:
https://github.com/abraunegg/onedrive/discussions/1689, https://github.com/abraunegg/onedrive/discussions/1801, https://github.com/abraunegg/onedrive/issues/1097, https://github.com/abraunegg/onedrive/issues/1908

However, I still confuse about why "fullscan" doesn't retry the file that return error before?
What will be the best practice to set up a mirror service under a internet while it would be so slow occasionally.

Best Reguards.

**Comment by abraunegg** [comment, marked as answer]
@1-1-2 
> However, I still confuse about why "fullscan" doesn't retry the file that return error before?

This is because the OneDrive API has 100% zero knowledge that your file failed to download. It has no server side tracking that there was a client side download issue.

Because of this, the /delta endpoint used to track the changes to send to you now no longer knows that this is a file that you need to download - because it thinks you downloaded the file sucessfully

There is no way to individually trigger a redownload outside of performing a --resync

> What will be the best practice to set up a mirror service under a internet while it would be so slow occasionally.

I would suggest:
* Investigate better Internet access to give you a better class of service as a starting point if you are trying to maintain a mirror or something
* Run the application via a cronjob rather than a systemd service - running 2 jobs where one is performing a 'sync' every x minutes or hours and a second process running once per day that uses --resync to ensure a trueup has occured. Schedule this trueup for when your Internet pipe is 'least congested'

**Reply by 1-1-2** [reply]
Thanks for explaining this. It may be a good idea to set up a service with a script to cover this. I will try this later.
